src/App.tsx

```diff
@@ -1,26 +1,254 @@
+        );
+      });
+  };
+
+  const handleTodoDelete = useCallback(
+    (todoId: number) => {
+      setIsSubmitting(true);
+      setLoadingTodosCount(current => [...current, todoId]);
+      deleteTodos(todoId)
+        .then(() => {
+          setTodos(currentTodos =>
+            currentTodos.filter(todo => todo.id !== todoId),
+          );
+        })
+        .catch(() => {
+          setErrorMessage(ErrorMessages.DELETING_ERROR);
+        })
+        .finally(() => {
+          setIsSubmitting(false);
+          setLoadingTodosCount(current =>
+            current.filter(deletingTodoId => todoId !== deletingTodoId),
+          );
+        });
+    },
+    [todos],
+  );
+  const handleCompletedTodosDeleted = useCallback(() => {
+    setIsSubmitting(true);
+
+    const todosForDeleting = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setLoadingTodosCount(current => [...current, ...todosForDeleting]);
+
+    Promise.all(
+      todosForDeleting.map(id =>
+        deleteTodos(id)
+          .then(() => {
+            setTodos(currentTodos =>
+              currentTodos.filter(todo => todo.id !== id),
+            );
+          })
+          .catch(() => {
+            setErrorMessage(ErrorMessages.DELETING_ERROR);
+          })
+          .finally(() => {
+            setLoadingTodosCount(current =>
+              current.filter(deletingTodoId => deletingTodoId !== id),
+            );
+          }),
+      ),
+    ).finally(() => setIsSubmitting(false));
+  }, [todos]);
+
+  const handleTodosToggle = useCallback(() => {
+    const hasIncompleteTodos = todos.some(todo => !todo.completed);
+
+    setTodos(currentTodos =>
+      currentTodos.map(todo => ({
+        ...todo,
+        completed: hasIncompleteTodos,
+      })),
+    );
+  }, [todos]);
+
+  const handleTodoToggle = useCallback(
+    (todoId: number, currentCompletedStatus: boolean) => {
+      setIsSubmitting(true);
+      changeTodo(todoId, { completed: !currentCompletedStatus })
+        .then(() => {
+          setTodos(currentTodos =>
+            currentTodos.map(todo =>
+              todo.id === todoId
+                ? { ...todo, completed: !todo.completed }
+                : todo,
+            ),
+          );
+        })
+        .catch(() => {
+          setErrorMessage(ErrorMessages.UPDATING_ERROR);
+        })
+        .finally(() => setIsSubmitting(false));
+    },
+    [todos],
+  );
+
+  const todosAfterFiltering = useMemo(
+    () => filteredTodos(todos, filter),
+    [todos, filter],
+  );
+
+  const completedTodos = useMemo(
+    () => todos.filter(todo => todo.completed),
+    [todos],
+  );
+
+  useEffect(() => {
+    const loadTodos = () => {
+      getTodos()
+        .then(loadedTodos => {
+          setTodos(loadedTodos);
+        })
+        .catch(error => {
+          setErrorMessage(ErrorMessages.LOADING_ERROR);
+          throw new Error(error);
+        });
+    };
```

Comment:
It's better to move the logic inside the useEffect hook to a separate function, as it improves code readability, reusability, and makes testing easier. It also helps keep useEffect focused on side effects.